src/Cake.Common/Tools/VSWhere/Product/VSWhereProduct.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Cake.Core;
+using Cake.Core.IO;
+using Cake.Core.Tooling;
+
+namespace Cake.Common.Tools.VSWhere.Product
+{
+    /// <summary>
+    /// The VSWhere tool that finds Visual Studio products.
+    /// </summary>
+    public sealed class VSWhereProduct : VSWhereTool<VSWhereProductSettings>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VSWhereProduct"/> class.
+        /// </summary>
+        /// <param name="fileSystem">The file system.</param>
+        /// <param name="environment">The environment.</param>
+        /// <param name="processRunner">The process runner.</param>
+        /// <param name="toolLocator">The tool locator.</param>
+        public VSWhereProduct(IFileSystem fileSystem, ICakeEnvironment environment, IProcessRunner processRunner,
+            IToolLocator toolLocator) : base(fileSystem, environment, processRunner, toolLocator)
+        {
+        }
+
+        /// <summary>
+        /// Finds one ore more products. Defaults to Community, Professional, and Enterprise.
+        /// Specify "*" by itself to search all product instances installed.
+        /// </summary>
+        /// <param name="settings">The settings.</param>
+        /// <returns>Installation paths for all instances.</returns>
+        public IEnumerable<FilePath> Products(VSWhereProductSettings settings)
+        {
+            if (settings == null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+            IEnumerable<string> installationPaths = new List<string>();
+            Run(settings, GetArguments(settings), new ProcessSettings { RedirectStandardOutput = true },
+                process => installationPaths = process.GetStandardOutput());
+            if (installationPaths.Any())
+            {
+                return installationPaths.Select(x => new FilePath(x));
+            }
+            return new List<FilePath>();
+        }
+
+        private ProcessArgumentBuilder GetArguments(VSWhereProductSettings settings)
+        {
+            var builder = new ProcessArgumentBuilder();
+
+            builder.Append("-products");
```

Comment:
@phillipsj  could you look at sorting this and I'll do a final review before merging this.